src/functions/env.ts

```diff
@@ -0,0 +1,238 @@
+import * as fs from "fs";
+import * as path from "path";
+
+import { FirebaseError } from "../error";
+import { logger } from "../logger";
+
+const RESERVED_KEYS = [
+  // Cloud Functions for Firebase
+  "FIREBASE_CONFIG",
+  // Cloud Functions - old runtimes:
+  //   https://cloud.google.com/functions/docs/env-var#nodejs_8_python_37_and_go_111
+  "ENTRY_POINT",
+  "GCP_PROJECT",
+  "GCLOUD_PROJECT",
+  "GOOGLE_CLOUD_PROJECT",
+  "FUNCTION_TRIGGER_TYPE",
+  "FUNCTION_NAME",
+  "FUNCTION_MEMORY_MB",
+  "FUNCTION_TIMEOUT_SEC",
+  "FUNCTION_IDENTITY",
+  "FUNCTION_REGION",
+  // Cloud Functions - new runtimes:
+  //   https://cloud.google.com/functions/docs/env-var#newer_runtimes
+  "FUNCTION_TARGET",
+  "FUNCTION_SIGNATURE_TYPE",
+  "K_SERVICE",
+  "K_REVISION",
+  "PORT",
+  // Cloud Run:
+  //   https://cloud.google.com/run/docs/reference/container-contract#env-vars
+  "K_CONFIGURATION",
+];
+
+// Regex to capture key, value pair in a dotenv file.
+// Inspired by:
+//   https://github.com/bkeepers/dotenv/blob/master/lib/dotenv/parser.rb
+// prettier-ignore
+const LINE_RE = new RegExp(
+  "^" +                    // begin line
+  "\\s*" +                 //   leading whitespaces
+  "(\\w+)" +               //   key
+  "\\s*=\\s*" +            //   separator (=)
+  "(" +                    //   begin optional value
+  `\\s*'(?:\\'|[^'])*'|` + //     single quoted or
+  `\\s*"(?:\\"|[^"])*"|` + //     double quoted or
+  "[^\\#\\r\\n]+" +        //     unquoted
+  ")?" +                   //   end optional value
+  "\\s*" +                 //   trailing whitespaces
+  "(?:\\#[^\\n]*)?" +      //   optional comment
+  "$",                     // end line
+  "gms"                    // flags: global, multiline, dotall
+);
+
+interface ParseResult {
+  envs: Record<string, string>;
+  errors: string[];
+}
+
+/**
+ * Parse contents of a dotenv file.
+ *
+ * Each line should contain key, value pairs, e.g.:
+ *
+ *   SERVICE_URL=https://example.com
+ *
+ * Values can be double quoted, e.g.:
+ *
+ *   SERVICE_URL="https://example.com"
+ *
+ * Double quoted values can include newlines, e.g.:
+ *
+ *   PUBLIC_KEY="-----BEGIN PUBLIC KEY-----\nABC\nEFG\n-----BEGIN PUBLIC KEY-----""
+ *
+ * or span multiple lines, e.g.:
+ *
+ *   PUBLIC_KEY="-----BEGIN PUBLIC KEY-----
+ *   ABC
+ *   EFG
+ *   -----BEGIN PUBLIC KEY-----"
+ *
+ * See test for more examples.
+ *
+ * @return {ParseResult} Result containing parsed key, value pairs and errored lines.
+ */
+export function parse(data: string): ParseResult {
+  const envs: Record<string, string> = {};
+  const errors: string[] = [];
+
+  data = data.replace(/\r\n?/, "\n"); // For Windows support.
+  let match;
+  while ((match = LINE_RE.exec(data))) {
+    let [, k, v] = match;
+    v = (v || "").trim();
+
+    let quotesMatch;
+    if ((quotesMatch = /^(["'])(.*)\1$/ms.exec(v)) != null) {
```

Comment:
Ugh.. TIL that you can actually inline regexp options after the trailing /. How is that not a parser can of worms? I wonder if `new RegExp("^([\"'])(.*)\1$", "ms")` is clearer, but I think it's all going to be horrific.